File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLRootQueryBranches.java
package net.nemerosa.ontrack.graphql.schema;

import graphql.schema.DataFetcher;
import graphql.schema.GraphQLFieldDefinition;
import net.nemerosa.ontrack.model.structure.Branch;
import net.nemerosa.ontrack.model.structure.ID;
import net.nemerosa.ontrack.model.structure.Project;
import net.nemerosa.ontrack.model.structure.StructureService;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.function.Predicate;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static graphql.Scalars.GraphQLInt;
import static graphql.Scalars.GraphQLString;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static net.nemerosa.ontrack.graphql.support.GraphqlUtils.checkArgList;
import static net.nemerosa.ontrack.graphql.support.GraphqlUtils.stdList;
import static org.apache.commons.lang3.StringUtils.isNotBlank;

@Component
public class GQLRootQueryBranches implements GQLRootQuery {

    private final StructureService structureService;
    private final GQLTypeBranch branch;
    private final GQLInputPropertyFilter propertyFilter;

    @Autowired
    public GQLRootQueryBranches(StructureService structureService, GQLTypeBranch branch, GQLInputPropertyFilter propertyFilter) {
        this.structureService = structureService;
        this.branch = branch;
        this.propertyFilter = propertyFilter;
    }

    @Override
    public GraphQLFieldDefinition getFieldDefinition() {
        return newFieldDefinition()
                .name("branches")
                .type(stdList(branch.getType()))
                .argument(
                        newArgument()
                                .name("id")
                                .description("ID of the branch to look for")
                                .type(GraphQLInt)
                                .build()
                )
                .argument(
                        newArgument()
                                .name("project")
                                .description("Name of the project the branch belongs to")
                                .type(GraphQLString)
                                .build()
                )
                .argument(
                        newArgument()
                                .name("name")
                                .description("Regular expression to match against the branch name")
                                .type(GraphQLString)
                                .build()
                )
                .argument(propertyFilter.asArgument())
                .dataFetcher(branchFetcher())
                .build();
    }

    private DataFetcher branchFetcher() {
        return environment -> {
            Integer id = environment.getArgument("id");
            String projectName = environment.getArgument("project");
            String name = environment.getArgument("name");
            Object propertyFilterArg = environment.getArgument(GQLInputPropertyFilter.ARGUMENT_NAME);
            // Per ID
            if (id != null) {
                checkArgList(environment, "id");
                return Collections.singletonList(
                        structureService.getBranch(ID.of(id))
                );
            }
            // Per project name, name or property filter
            else if (isNotBlank(projectName) || isNotBlank(name) || propertyFilterArg != null) {

                // Project filter
                Predicate<Project> projectFilter = p -> true;
                if (isNotBlank(projectName)) {
                    projectFilter = projectFilter.and(project -> StringUtils.equals(projectName, project.getName()));
                }

                // Branch filter
                Predicate<Branch> branchFilter = b -> true;
                if (isNotBlank(name)) {
                    Pattern pattern = Pattern.compile(name);
                    branchFilter = branchFilter.and(b -> pattern.matcher(b.getName()).matches());
                }

                // Property filter?
                if (propertyFilterArg != null) {
                    GQLInputPropertyFilter.PropertyFilter filterObject = propertyFilter.convert(propertyFilterArg);
                    if (filterObject != null && StringUtils.isNotBlank(filterObject.getType())) {
                        branchFilter = branchFilter.and(propertyFilter.getFilter(filterObject));
                    }
                }

                // Gets the list of authorised projects
                return structureService.getProjectList().stream()
                        // Filter on the project
                        .filter(projectFilter)
                        // Gets the list of branches
                        .flatMap(project -> structureService.getBranchesForProject(project.getId()).stream())
                        // Filter on the branch
                        .filter(branchFilter)
                        // OK
                        .collect(Collectors.toList());
            }
            // No result to return
            else {
                return Collections.emptyList();
            }
        };
    }

}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLRootQueryBranches.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars.*
import graphql.schema.DataFetcher
import graphql.schema.GraphQLArgument.newArgument
import graphql.schema.GraphQLFieldDefinition
import graphql.schema.GraphQLFieldDefinition.newFieldDefinition
import net.nemerosa.ontrack.common.and
import net.nemerosa.ontrack.common.getOrNull
import net.nemerosa.ontrack.graphql.support.GraphqlUtils.checkArgList
import net.nemerosa.ontrack.graphql.support.GraphqlUtils.stdList
import net.nemerosa.ontrack.model.structure.*
import org.springframework.stereotype.Component

const val GRAPHQL_BRANCHES_FAVORITE_ARG = "favourite"

@Component
class GQLRootQueryBranches(
        private val structureService: StructureService,
        private val branch: GQLTypeBranch,
        private val propertyFilter: GQLInputPropertyFilter,
        private val branchFavouriteService: BranchFavouriteService
) : GQLRootQuery {

    override fun getFieldDefinition(): GraphQLFieldDefinition {
        return newFieldDefinition()
                .name("branches")
                .type(stdList(branch.typeRef))
                .argument(
                        newArgument()
                                .name("id")
                                .description("ID of the branch to look for")
                                .type(GraphQLInt)
                                .build()
                )
                .argument(
                        newArgument()
                                .name("project")
                                .description("Name of the project the branch belongs to")
                                .type(GraphQLString)
                                .build()
                )
                .argument(
                        newArgument()
                                .name("name")
                                .description("Regular expression to match against the branch name")
                                .type(GraphQLString)
                                .build()
                )
                .argument {
                    it.name(GRAPHQL_BRANCHES_FAVORITE_ARG)
                            .description("Keeps only branches listed as favourute")
                            .type(GraphQLBoolean)
                }
                .argument(propertyFilter.asArgument())
                .dataFetcher(branchFetcher())
                .build()
    }

    private fun branchFetcher(): DataFetcher<List<Branch>> = DataFetcher { environment ->
        val id: Int? = environment.getArgument("id")
        val projectName: String? = environment.getArgument("project")
        val name: String? = environment.getArgument("name")
        val favourite: Boolean? = environment.getArgument(GRAPHQL_BRANCHES_FAVORITE_ARG)
        val propertyFilterArg: Any? = environment.getArgument(GQLInputPropertyFilter.ARGUMENT_NAME)

        // Project filter
        val projectFilter: (Project) -> Boolean = if (projectName.isNullOrBlank()) {
            { true }
        } else {
            { project -> project.name == projectName }
        }

        // Branch name filter
        val regex = name?.toRegex()
        val branchNameFilter: (Branch) -> Boolean = if (regex == null) {
            { true }
        } else {
            { branch -> regex.matches(branch.name) }
        }

        // Property filter?
        val branchPropertyFilter: (Branch) -> Boolean = if (propertyFilterArg != null) {
            val filterObject: PropertyFilter? = propertyFilter.convert(propertyFilterArg)
            val type = filterObject?.type
            if (filterObject != null && type != null && type.isNotBlank()) {
                { b: Branch -> propertyFilter.getFilter(filterObject).test(b) }
            } else {
                { true }
            }
        } else {
            { true }
        }

        // Branch filter
        val branchFilter = branchNameFilter and branchPropertyFilter

        // Per ID
        if (id != null) {
            checkArgList(environment, "id")
            listOf(structureService.getBranch(ID.of(id)))
        } else if (favourite == true) {
            // Gets the list of favourite branches
            val branches = branchFavouriteService.getFavouriteBranches()
            // Filter them with project, name & property
            branches.filter {
                projectFilter(it.project) && branchFilter(it)
            }
        } else if (!projectName.isNullOrBlank()) {
            // Gets the project
            val project = structureService.findProjectByName(projectName).getOrNull()
            // Gets the branches for this project
            val branches = project?.let { structureService.getBranchesForProject(project.id) } ?: emptyList()
            // Filter them with name & property
            branches.filter(branchFilter)
        } else {
            // Gets the projects
            structureService.projectList.asSequence()
                    // Filter on the project
                    .filter(projectFilter)
                    // Gets the list of branches
                    .flatMap { project -> structureService.getBranchesForProject(project.id).asSequence() }
                    // Filter on the branch
                    .filter(branchFilter)
                    // OK
                    .toList()
        }
    }

}
